// ffplay/lib/ffplay.js
const ffmpegPath = require("@ffmpeg-installer/ffmpeg").path;
var spawn = require("child_process").spawn,
	EventEmitter = require("events").EventEmitter;

class FFplay extends EventEmitter {
	constructor(file, opts) {
		this.pausing = false
		this.running = false

		opts = opts || ["-nodisp", "-autoexit"];
		opts.unshift(file);
		this.proc = spawn(ffmpegPath.replace('ffmpeg.exe', 'ffplay.exe'), opts, { stdio: "ignore" });
		this.ef = function () {
			this.proc.kill();
		}.bind(this);
		process.on("exit", this.ef);
		this.proc.on("exit", function () {
			if (this.running) {
				this.running = false;
				process.removeListener("exit", this.ef);
				if (!this.manualStop) {
					setImmediate(function () { this.emit("stopped"); }.bind(this));
				}
			}
		}.bind(this));
		this.running = true;
	}
	pause() {
		if (!this.paused) {
			this.proc.kill("SIGSTOP");
			this.paused = true;
			this.emit("paused");
		}
	}
	resume() {
		if (this.paused) {
			this.proc.kill("SIGCONT");
			this.paused = false;
			this.emit("resumed");
		}
	}
	stop() {
		this.manualStop = true;
		this.proc.kill("SIGKILL");
	}
}

module.exports = FFplay;
